test/generic.jl

```diff
@@ -5,34 +5,38 @@ using ForwardDiff: Dual
 function check_rmath(fname, statsfun, rmathfun, params, aname, a, isprob, rtol)
     v = @inferred(rmathfun(params..., a))
     rv = @inferred(statsfun(params..., Dual(a))).value
+    @test v isa float(Base.promote_typeof(params..., a))
+    @test rv isa float(Base.promote_typeof(params..., a))
     if isprob
-        rd = abs(v / rv - 1.0)
-        if rd > rtol
-            error("$fname deviates too much from Rmath at " *
-                "params = $params, $aname = $a:\n" *
-                "  v = $v (rv = $rv)\n  |v/rv - 1| = $rd > $rtol.")
-        end
+        @test v ≈ rv rtol=rtol nans=true
     else
-        τ = (1.0 + abs(rv)) * rtol
-        ad = abs(v - rv)
-        if ad > τ
-            error("$fname deviates too much from Rmath at " *
-                "params = $params, $aname = $a:\n" *
-                "  v = $v (rv = $rv)\n  |v - rv| = $ad > $τ.")
-        end
+        @test v ≈ rv atol=rtol rtol=rtol nans=true
     end
 end
 
-function genericcomp(basename, params, X::AbstractArray, rtol=100eps(float(one(eltype(X)))))
+function genericcomp(basename, params, X::AbstractArray)
+    # compute default tolerance:
+    # has to take into account `params` as well since otherwise e.g. `X::Array{<:Rational}`
+    # always uses a tolerance based on `eps(one(Float64))` even when parameters are of type
+    # Float32
+    # eps^(7//8) means requiring equality of about 7/8 of the significant digits
+    # Corresponds to tolerances of ~2e-14 (Float64), ~9f-7 (Float32) and ~0.002 (Float16)
+    rtol = eps(float(one(promote_type(Base.promote_typeof(params...), eltype(X)))))^(7//8)
```

Comment:
This is even a tiny bit smaller than the current tolerance (currently `100eps(Float64)  = 2.220446049250313e-14` whereas `eps(Float64)^(7//8) = 2.0097183471152322e-14`). But exponentiation seems to be more useful/interpretable generally, in particular when working with other floating point types. So I think we should change the default tolerance in rmath.jl as well, since there we actually work with other types such as `Float16` and `Float32`.